design/httpproxy-conflict.md

```diff
@@ -0,0 +1,125 @@
+Configurations that are in an Error or Warning status, will serve HTTP error codes or not be added to configuration passed down to Envoy.
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress that reduces annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to ResourceOwners (See #Personas section below) via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+It's difficult to find errors and block them from breaking valid configurations since Contour currently processes objects after they are committed to the API server.
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`).
+When processing HTTPProxy objects, some problems should not stop the rest of the object from being processed.
+To distinguish between fatal-for-processing and not, we propose calling the former `Errors`, and the latter `Warnings`.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, only the first is shown to the user.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes or even the entire virtual host.
+
+When processing HTTPProxy objects, it turns out some problems should not stop the rest of the object from being processed.
+To distinguish between fatal-for-processing and not, we propose calling the former Errors, and the latter Warnings.
+
+### Personas
+
+There are two main personas that interact with this conflict design document:
+
+- User: A request to an application, routing through Envoy looking to access whatever resource an application running in Kubernetes should serve. This user is not aware of how the application is deployed or that it uses Kubernetes or Contour.
+- ResourceOwner: Someone to manages an HTTPProxy resource inside the Kubernetes cluster configuring the Route/Service combination to allow users of the application to access.  
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+- Define and distinguish between fatal and non-fatal errors
+
+## Non Goals
+- Change how Contour processes objects today from informers set against the Kubernetes API
+
+## High-Level Design
+Contour will set the error or warning condition when a problem is encountered, but do its best to still serve valid configurations.
+
+In general, Contour will set an object to be status `Error` if the request response to the `User` is changed from what is configured.
```

Comment:
L28 and L47 seem to give different definitions for `Error` - L28 says it means "fatal-for-processing", i.e. no part of the proxy will be programmed with Envoy, whereas L47 says it means the response to the user is changed. Can we clear that up? Do *some* errors mean no part of the proxy is programmed with Envoy, whereas *other* errors can indicate partial programming? Or should the latter always correspond with a `Warning`? I think we need to be super-clear about this for users' sake.

Reply:
I agree that we should be super clear. My original idea about this was closer to the latter - any entry in the `Errors` slice in the `DetailedCondition` implies a fatal-for-processing, unrecoverable error that means that no config from that HTTPProxy will end up in Envoy.

When a HTTPProxy with an `Error` is included in another, the includ**ing** HTTPProxy must have a `Warning` at least, unless that would mean that there are no valid Routes, in which case the includ**ing** HTTPProxy would also have an `Error`.

Reply:
> unrecoverable error that means that no config from that HTTPProxy will end up in Envoy

This is 100% NOT my goal in this design. One HTTPProxy can have errors (say an invalid service), but still serve routes to other proxies.  

I removed line 28 and moved line 47 it up to where line 28 was.

Any entry in the Error slice tells that the response is changed (e.g. 503, etc), that's it.

Reply:
So how do we tell the difference then between "you made a big mistake in your HTTPProxy" and "Oh, turns out your service doesn't exist"? I *thought* that was the distinction between `Error` and `Warning`?

Reply:
To put this another way: I thought that `Errors` basically mapped to the old, non-condition `Valid` field. I thought that entries in the `Errors` field would be the same as the thing being `Invalid`.

Reply:
The goal of this design is to make sure users applications serve traffic. 

For example if a root proxy resource has 2 paths: `/foo` & `/bar` and the service for `/foo` is missing, today both paths are invalid and won't serve traffic. This design aims to allow `/bar` to still serve traffic even if the service for `/foo` is missing. 

Since the service is missing, this design changes how Contour handles this scenario, which means that any requests to `/foo` will return an HTTP 503 and requests to `/bar` will still route properly. 

Since the request is changed from it's original configuration (meaning the service is missing and now returns a 503), that's an Error. Any request to the path `/foo` should route to a service, but doesn't, they get a 503 response. 

A different scenario where the configured path is not changed, is a warning, which says, "hey we tried to configure this, something's wrong, but your user will still get to the place configured" (at best). 

See: https://github.com/projectcontour/contour/pull/3145/files#diff-dc07f1241f06c945856ac6f7fc155659c36976138500362c6955ccdf02adbc70R34-R36

Reply:
I understand that, so do you mean that you feel there's no distinction between "hey, most of this config worked" and "you messed up your FQDN, nothing did"?

Because right now, both of those things are Errors. I feel like the latter is a more severe class of Error, that we should treat differently.

I agree that we should be setting things to 503 when the service exists, and that that's why we started this. I get the rule about needing to tell the user that things are different to what they asked for, but I think there's a difference in degree between "one of your endpoints is busted" and "none of this config will work at all". If we use `Error` to represent the former, that leaves us with nothing for the latter.

Also, what about when it's a combination of headers that we can't satisfy? That is, the headers produce an empty set? Is that a warning or an error? Feels like a warning to me, but I'm not sure if I'm right here.

Reply:
> I understand that, so do you mean that you feel there's no distinction between "hey, most of this config worked" and "you messed up your FQDN, nothing did"?

What do you suggest then? I feel like calling both of those "Errors" is fine. Are you proposing we add a "SevereErrors" set to the conditions? 

> Also, what about when it's a combination of headers that we can't satisfy? That is, the headers produce an empty set? Is that a warning or an error? Feels like a warning to me, but I'm not sure if I'm right here.

Warning

Reply:
Some of this is what @skriss brought up about informing the user that the config was rejected or only served a partial config. Short of explaining in the error messages, I don't think we have a good way to explain that to users.

Reply:
We talked about this at length in the community office hours today. Here's where I think we ended up:

- both "fatal errors" (e.g. an invalid FQDN) and "non-fatal errors" (e.g. a route references a non-existent service) should be reported as `Errors` under the `Valid` Condition.
- in the case of "non-fatal errors", Contour should continue to process the `HTTPProxy` and attempt to program as much of it as possible with Envoy. This is particularly beneficial for when an `HTTPProxy` already exists and is valid, and has a bad edit made to it (e.g. to a single route). We don't want to stop serving traffic to other unaffected routes if we can avoid it.
- the result of the above two points is that an `HTTPProxy` with a `Valid: false` condition _may_ be partially programmed in Envoy, or may not be programmed at all in Envoy
- **it's not essential to differentiate between those cases**; either way, the user will be told something is wrong with the HTTPProxy via the `Valid: false` condition, and they will want to remediate it.

@youngnick, originally I was on the other side of the last bullet, i.e. I thought we needed to differentiate between partial programming and no programming, but I no longer think it's essential since either way, the user is told about the problem and will want to fix it. I *do* think that in the long run, we probably want to look at not just having a single, binary `Valid` condition, but instead having more top-level conditions that, in combination, could describe these shades of gray better. However, (a) I don't think we need to do that now, and (b) that's probably considered a breaking change that requires a v2 anyway.

Thoughts?